mlos_bench/mlos_bench/storage/sql_experiment.py

```diff
@@ -0,0 +1,178 @@
+
+    def load(self, opt_target: Optional[str] = None) -> Tuple[List[dict], List[float]]:
+        configs = []
+        scores = []
+        with self._engine.connect() as conn:
+            cur_trials = conn.execute(
+                text("""
+                SELECT t.trial_id, t.config_id, r.metric_value
+                FROM trial AS t
+                JOIN trial_result AS r ON (t.exp_id = r.exp_id AND t.trial_id = r.trial_id)
+                WHERE t.exp_id = :exp_id AND t.status = 'SUCCEEDED' AND r.metric_id = :metric_id
+                ORDER BY t.trial_id ASC
+                """),
+                {
+                    "exp_id": self._experiment_id,
+                    "metric_id": opt_target or self._opt_target
+                }
+            )
+            for trial in cur_trials.fetchall():
+                tunables = self._get_params(
+                    conn, self._schema.config_param, config_id=trial.config_id)
+                configs.append(tunables)
+                scores.append(float(trial.metric_value))
+            return (configs, scores)
+
+    @staticmethod
+    def _get_params(conn, table, **kwargs) -> dict:
+        cur_params = conn.execute(table.select().where(*[
+            column(key) == val for (key, val) in kwargs.items()]))
+        return {row.param_id: row.param_value for row in cur_params.fetchall()}
+
+    @staticmethod
+    def _save_params(conn, table, params: dict, **kwargs):
+        conn.execute(table.insert(), [
+            {
+                **kwargs,
+                "param_id": key,
+                "param_value": None if val is None else str(val)
+            }
+            for (key, val) in params.items()
+        ])
+
+    def pending(self):
+        _LOG.info("Retrieve pending trials for: %s", self._experiment_id)
+        with self._engine.connect() as conn:
+            cur_trials = conn.execute(self._schema.trial.select().where(
+                column("exp_id") == self._experiment_id,
+                column("ts_end").is_(None)
+            ))
+            for trial in cur_trials.fetchall():
+                tunables = self._get_params(
+                    conn, self._schema.config_param,
+                    config_id=trial.config_id)
+                config = self._get_params(
+                    conn, self._schema.trial_param,
+                    exp_id=self._experiment_id, trial_id=trial.trial_id)
+                # Reset .is_updated flag after the assignment:
+                yield Trial(self._engine,
+                            self._tunables.copy().assign(tunables).reset(),
+                            self._experiment_id, trial.trial_id, self._opt_target, config)
+
+    def _get_config_id(self, conn, tunables: TunableGroups) -> int:
+        """
+        Get the config ID for the given tunables. If the config does not exist,
+        create a new record for it.
+        """
+        config_hash = hashlib.sha256(str(tunables).encode('utf-8')).hexdigest()
```

Comment:
`... str(tunable) ...`

Do we ensure that this is always coming in a canonical form?

e.g. `json.dumps(indent=None, sort_keys=True)` or something like that.

Reply:
yes, `TunableGroups.__repr__()` produces a canonical form, although it drops some information (e.g., cost). Later we can implement both `.__str__()` and `.__repr__()` if we want both canonical and more user-friendly string representations.

Reply:
Sounds good.  Maybe we can still add a test for it:

e.g.
```python
# Python uses ordered dicts.
config_a = json.loads('{ "a": 1, "b": 2 }')
config_b = json.loads('{ "b": 2, "a": 1 }')
assert hash(config_a) == hash(config_b)
```

Reply:
ok I've added some unit tests for this, too. I had to update the data in other unit tests, though, because I have touched the TunableGroups fixture and now all optimizers produce new results :)